File: Core/src/main/java/com/inverce/mod/core/configuration/Value.java
package com.inverce.mod.core.configuration;

import android.support.annotation.NonNull;

import com.inverce.mod.core.configuration.extended.BoxedValue;
import com.inverce.mod.core.functional.IConsumer;
import com.inverce.mod.core.functional.IPredicate;
import com.inverce.mod.core.functional.ISupplier;

@SuppressWarnings({"unused", "WeakerAccess"})
public class Value<T> extends ReadOnlyValue<T> {
    protected IPredicate<T> validator;
    protected IConsumer<T> setter;
    protected ChangeValueHandler<T> changeValueHandler;

    protected Value() {
        changeValueHandler = new ChangeValueHandler<>();
    }

    public Value(T value) {
        this(value, p -> true);
    }

    public Value(T value, IPredicate<T> validator) {
        this();
        BoxedValue<T> box = new BoxedValue<>(value);
        setSetter(box::setValue);
        setGetter(box::getValue);
        setValidator(validator);
        set(value);
    }

    public Value(ISupplier<T> supplier, IConsumer<T> setter) {
        this(supplier, setter, p -> true);
    }

    public Value(ISupplier<T> supplier, IConsumer<T> setter, IPredicate<T> validator) {
        super(supplier);
        this.changeValueHandler = new ChangeValueHandler<>();
        this.validator = validator;
        this.setter = setter;
    }

    @NonNull
    public ValueChangedEvent<ValueChanged<T>> changeValueEvent() {
        return changeValueHandler;
    }

    public boolean set(T e) {
        if (validator.test(e)) {
            setter.accept(e);
            changeValueHandler.postNewValue(this, e);
            return true;
        }
        return false;
    }

    protected void setValidator(IPredicate<T> validator) {
        this.validator = validator;
    }

    protected void setSetter(IConsumer<T> setter) {
        this.setter = setter;
    }

    public ReadOnlyValue<T> asReadOnly() {
        return new ReadOnlyValue<>(this);
    }

}

File: Core/src/main/java/com/inverce/mod/core/configuration/Value.kt
package com.inverce.mod.core.configuration

import kotlin.reflect.KProperty

typealias ValueChanged<T> = (preference: Value<T>, value: T) -> Unit

interface ValueChangedEvent<T> {
    fun addListener(listener: T)
    fun removeListener(listener: T)
    fun clear()
}

open class ReadOnlyValue<T> protected constructor() {
    protected lateinit var getter: () -> T

    constructor(value: T) : this({ value })
    constructor(value: Value<T>) : this({ value.getValue() })
    constructor(getter: () -> T) : this() {
        this.getter = getter
    }

    open fun getValue(): T = getter()
    operator fun getValue(t: Any?, property: KProperty<*>): T = getValue()
    override fun toString(): String {
        return "Value: " + (getValue() ?: "null")
    }
}

open class Value<T> protected constructor(protected var setter: (T) -> Unit = { }, var validator: (T) -> Boolean = { true }) : ReadOnlyValue<T>() {
    val onChanged = listOf<(T) -> Unit>()

    constructor(getter: () -> T, setter: (T) -> Unit, validator: (T) -> Boolean = { true }) : this(setter, validator) {
        this.getter = getter
    }

    constructor(value: T, validator: (T) -> Boolean = { true }) : this() {
        val box = arrayListOf(value)
        setter = { box[0] = it }
        getter = { box[0] }
        this.validator = validator
    }

    open fun setValue(e: T) = when (validator(e)) {
        true -> {
            setter(e)
            onChanged.onEach { it(e) }
            true
        }
        else -> false
    }

    operator fun setValue(t: Any?, property: KProperty<*>, value: T) = setValue(value)

    fun asReadOnly(): ReadOnlyValue<T> = ReadOnlyValue(this)

}